### DISCUSSION TITLE
Websockets: how to detect the final frame in a message with curl_ws_recv()?

### DISCUSSION THREAD

**Discussion by anovicemaster** [curl/curl#14249] [Q&A] [answered]
I found this exact question on the mailing list from a year ago, but without a definitive answer/solution so I'm posting here. Link to the original question: https://curl.se/mail/lib-2023-07/0007.html

### Question
Suppose I'm connected to a websocket server in `CURLOPT_CONNECT_ONLY` mode. The server sends each message in variable number of frames (can be 1, 2, 3 or more frames, no guarantee). My program is running curl_ws_recv() in a loop to concatenate the frames together into a complete message. But how do I know when to stop the loop, (i.e. we've read the final frame in the current message, and the next frame received will be a new message)?

### My Environment
I'm running curl-8.7.1 on aarch64, Oracle Linux 9

### What I've Done
The first thing I tried is to check the `CURLWS_CONT` flag on each iteration. Based on the documentation for this flag: 
> This is not the final fragment of the message, it implies that there is another fragment coming as part of the same message.

I thought the `CURLWS_CONT` bit would be set on every frame received except the **final** frame, and I could write my loop like so:
``` c
const struct curl_ws_frame* meta;
size_t chunk_size;

do {
    CURLcode result = curl_ws_recv(curl, cursor, avail_size, &chunk_size, &meta);
    // check error, move cursor, reduce avail_size, do other stuff, etc.
} while (meta->flags & CURLWS_CONT); // if CURLWS_CONT is set, continue loop. Otherwise, exit.
```
But this does not work. It either exits early (partial message) or late (incorrect, with extra parts of the next message). When I print out `meta->flags` after receiving each frame, I see the `CURLWS_CONT` flag set on every frame except the **first** frame. This makes me think `meta->flags` is simply mirroring the actual opcodes, because the behavior here agrees with the RFC definition:
>A fragmented message consists of a single frame with the FIN bit clear and an opcode other than 0, followed by zero or more frames with the FIN bit clear and the opcode set to 0, and terminated by a single frame with the FIN bit set and an opcode of 0.

If that's the case, then we'll need an additional flag/bit somewhere that mirrors the FIN bit in order to determine the end of a message. Looking everywhere through documentation and relevant header files I could not find an exposed bit mask for the FIN bit. I also checked `meta->bytesleft` just in case but it's 0 on every frame so not exactly helpful.

I eventually tried a quick hack in ws.c to preserve the FIN bit, re-compiled libcurl and my program was able to return correct messages by checking the FIN bit instead. There surely has to be a better way, or maybe I'm just missing something obvious?

Appreciate any help and suggestions!

**Comment by jay** [comment, marked as answer]
> I thought the `CURLWS_CONT` bit would be set on every frame received except the **final** frame

AFAICT you have to write it like `meta->bytesleft || (meta->flags & CURLWS_CONT)`. The reason it doesn't work though is possibly a bug. Websockets support is still experimental. I have a PR #14214 that improves on the websocket example where I did exactly that but I think some is still up in the air, like can some messages interrupt others. We don't have a websocket testing framework to similuate all these things. Is it a public server and do you have a way to reproduce? I used echo.websocket.org when I was writing the example to test meta->bytesleft but there was no `CURLWS_CONT` IIRC

**Reply by anovicemaster** [reply]
Thank you for the response Jay. I see your point, although I run into the same issue using `meta->bytesleft || (meta->flags & CURLWS_CONT)`. Unfortunately it's not a public server.

As far as I've tested and understood, `meta->bytesleft` addresses the multi-fragment scenario within a single frame. If a frame is not being fragmented internally by libcurl, which is largely the case for my program, `meta->bytesleft` would just be 0 and that makes sense. I've previously tested the `meta->bytesleft` part both ways (with and without fragmentation) and it's working as intended. I believe your examples and tests also cover this part. 

Meanwhile, `meta->flags & CURLWS_CONT` is supposed to address the multi-frame scenario within a single message, but it doesn't. For the same reasons I detailed in my original post. I certainly get it if this is a bug given WS is experimental. For the most part, I'm just trying to make sure I'm not missing an obvious solution or grossly misunderstanding the intended usage.
